Code/6-wire-version/src/main.cpp

```diff
@@ -21,78 +21,113 @@ void setup() {
-  periodicTimer.attach(60, periodicTimerCallback);  //things to do once in a while
-  LittleFS.begin(); // needs to be loaded here for reading the wifi.json
+
+  // check things every 60 seconds
+  periodicTimer.attach(60, []{ periodicTimerFlag = true; });
+  
+  // update webpage every 2 seconds. (will also be updated on state changes)
+  updateWSTimer.attach(2.0, []{ sendWSFlag = true; });
+
+  // update MQTT every 10 minutes. (will also be updated on state changes)
+  updateMqttTimer.attach(600, []{ sendMQTTFlag = true; });
+
+  // needs to be loaded here for reading the wifi.json
+  LittleFS.begin();
   loadWifi();
   startWiFi();
   startNTP();
   startOTA();
-  startServer();
+  startHttpServer();
   startWebSocket();
   startMQTT();
+
   bwc.print(WiFi.localIP().toString());
-  Serial.println(F("\nEnd of setup()"));
+  Serial.println(F("End of setup()"));
 }
 
-void loop() {
-  bwc.loop();                     // Fiddle with the pump computer
-  if(WiFi.status() == WL_CONNECTED){
-    webSocket.loop();             // check for websocket events
-    server.handleClient();        // run the server
-    ArduinoOTA.handle();          // listen for OTA events
-      
-    if (enableMqtt){
-      if(!MQTTclient.loop()){
+void loop()
+{
+  // Fiddle with the pump computer
+  bwc.loop();
+
+  // do things only when WiFi is connected
+  if (WiFi.status() == WL_CONNECTED)
+  {
+    // listen for websocket events
+    webSocket.loop();
+    // listen for webserver events
+    server.handleClient();
+    // listen for OTA events
+    ArduinoOTA.handle();
+    
+    // MQTT
+    if (enableMqtt)
+    {
+      if (!MQTTclient.loop())
+      {
         MQTT_Connect();
-      } else {
+      }
+      else
+      {
         String msg = bwc.getButtonName();
-        //publish pretty button name if display button is pressed (or NOBTN if released)
-        if(!msg.equals(prevButtonName)) {
+        // publish pretty button name if display button is pressed (or NOBTN if released)
+        if (!msg.equals(prevButtonName))
+        {
           MQTTclient.publish((String(mqttBaseTopic) + "/button").c_str(), String(msg).c_str(), true);
           prevButtonName = msg;
         }
-        if (bwc.newData()) sendMQTT();
+
+        if (bwc.newData())
+        {
+          sendMQTT();
+        }
+        else if (sendMQTTFlag)
+        {
+          sendMQTTFlag = false;
+          sendMQTT();
+        }
       }
     }
     
-    if (bwc.newData()) {
+    // web socket
+    if (bwc.newData())
+    {
       sendWS();
     }
-    if (sendWSFlag) {
+    else if (sendWSFlag)
+    {
       sendWSFlag = false;
       sendWS();
     }
-    if (sendMQTTFlag && enableMqtt) {
-      sendMQTTFlag = false;
-      sendMQTT();
-    }
   }
 
-  if(periodicTimerFlag){
+  if (periodicTimerFlag)
+  {
     periodicTimerFlag = false;
-    if (WiFi.status() != WL_CONNECTED) {
+    if (WiFi.status() != WL_CONNECTED)
+    {
```

Comment:
@visualapproach do we need some reconnect here..?

Reply:
Not as is. ESP OS is doing that every second.

Reply:
i think i will test a bit around and then come back to this